sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/bigquery/BigQueryIO.java

```diff
@@ -319,149 +328,193 @@ public static Read read() {
         .build();
   }
 
-  /** Implementation of {@link #read}. */
-  @AutoValue
-  public abstract static class Read extends PTransform<PBegin, PCollection<TableRow>> {
-    @Nullable abstract ValueProvider<String> getJsonTableRef();
-    @Nullable abstract ValueProvider<String> getQuery();
-    abstract boolean getValidate();
-    @Nullable abstract Boolean getFlattenResults();
-    @Nullable abstract Boolean getUseLegacySql();
+  /**
+   * A {@link PTransform} that reads from a BigQuery table and returns a
+   * {@link PCollection} of records transformed by a user defined conversion function.
+   */
+  public static <T>ReadGenericRecords<T> readRecords() {
+    return new AutoValue_BigQueryIO_ReadGenericRecords.Builder<T>()
+        .setValidate(true)
+        .setWithTemplateCompatibility(false)
+        .setBigQueryServices(new BigQueryServicesImpl())
+        .build();
+  }
 
-    abstract Boolean getWithTemplateCompatibility();
+  private static class TableSchemaFunction
+      implements Serializable, Function<String, TableSchema> {
```

Comment:
This should use a SerializableFunction

Reply:
can do, I just moved this out of BigQuerySourceBase though :P

Reply:
jk, its not a SerializableFunction apparently because it needs to extend guava's `Function` for `Suppliers.compse` below.